src/Internal/BalanceTx/ExUnitsAndMinFee.purs

```diff
@@ -80,28 +81,33 @@ evalTxExecutionUnits
   :: Transaction
   -> BalanceTxM Ogmios.TxEvaluationResult
 evalTxExecutionUnits tx = do
-  queryHandle <- liftContract getQueryHandle
-  additionalUtxos <- getOgmiosAdditionalUtxoSet
-  evalResult <-
-    unwrap <$> liftContract
-      (liftAff $ queryHandle.evaluateTx tx additionalUtxos)
-
-  case evalResult of
-    Right a -> pure a
-    Left evalFailure | tx ^. _isValid ->
-      throwError $ ExUnitsEvaluationFailed tx evalFailure
-    Left _ -> pure $ wrap Map.empty
+  networkId <- askNetworkId
+  additionalUtxos <-
+    fromPlutusUtxoMap networkId <$> asksConstraints Constraints._additionalUtxos
+  worker additionalUtxos
   where
-  getOgmiosAdditionalUtxoSet :: BalanceTxM Ogmios.AdditionalUtxoSet
-  getOgmiosAdditionalUtxoSet = do
-    networkId <- askNetworkId
-    additionalUtxos <-
-      asksConstraints Constraints._additionalUtxos
-        <#> fromPlutusUtxoMap networkId
-    pure $ wrap $ Map.fromFoldable
-      ( bimap transactionInputToTxOutRef transactionOutputToOgmiosTxOut
-          <$> (Map.toUnfoldable :: _ -> Array _) additionalUtxos
-      )
+  worker :: UtxoMap -> BalanceTxM Ogmios.TxEvaluationResult
+  worker additionalUtxos = do
+    queryHandle <- liftContract getQueryHandle
+    evalResult <-
+      unwrap <$> liftContract
+        (liftAff $ queryHandle.evaluateTx tx ogmiosAdditionalUtxos)
+    case evalResult of
+      Right a -> pure a
+      Left (Ogmios.AdditionalUtxoOverlap overlappingUtxos) ->
+        -- Remove overlapping additional utxos and retry evaluation:
+        worker $ Map.filterKeys (flip Array.notElem overlappingUtxos)
+          additionalUtxos
```

Comment:
`ogmiosAdditionalUtxos` will be recomputed on every recursive call. Can we compute it once outside of worker? `additionalUtxos` would not be needed at all then